whats-that-typeclass-functor/article.md

```diff
@@ -0,0 +1,467 @@
+
+After reading this article, you will:
+
+- know what Functor is in Haskell;
+- understand how to define and use your own Functor instances;
+- see why functors are useful and where they could be used;
+- be able to consolidate your knowledge with exercises.
+
+## How to generalize `map`
+
+Look at the type signature of `map`:
+
+```haskell
+map :: (a -> b) -> [a] -> [b]
+```
+
+It takes a function `a -> b`. Then, it uses that function to turn its second argument (of type `[a]`) into a list of `b` — `[b]`.
+
+Now imagine we’re using the same type of function — `a -> b` — to change the contents of  `Maybe`. For now, it doesn’t matter how exactly we’d do it. What type would such a `map'` function have?
+
+<details>
+<summary>The type of <code>map'</code> modifying <code>Maybe</code></summary>
+
+```haskell
+map' :: (a -> b) -> Maybe a -> Maybe b
+```
+<hr>
+</details>
+
+Finally, we can write a `map'` for an arbitrary data type `f` with one type argument `a` — `f a`. By analogy, can you figure out how it would look like?
+
+<details>
+<summary>The type of <code>map'</code> modifying <code>f</code></summary>
+
+```haskell
+map' :: (a -> b) -> f a -> f b
+```
+<hr>
+</details>
+
+Great! This `map'` is called `fmap` or `(<$>)` in Haskell.
+
+Here’s how it works: `fmap` takes an  `a -> b` function and an `f a` data type (`a` wrapped in a context `f`), then the function is applied to what’s inside the context, and finally, a value of type `b` wrapped in `f` is returned.
+
+Therefore, the value changes, but the context remains the same.
+
+Here are a few examples:
+
+```haskell
+-- apply `reverse :: String -> String` function
+-- to the list of `String`s
+-- to get the list of `String`s
+> fmap reverse ["abc", "def", "ghi"]
+["cba","fed","ihg"]
+> fmap reverse []
+[]
+
+-- apply `show :: Int -> String` function
+-- to the list of `Int`s
+-- to get the list of `String`s
+> fmap show [1..3]
+["1","2","3"]
+> fmap show []
+[]
+
+-- apply `(+1) :: Int -> Int` function
+-- to `Maybe Int`
+-- to get `Maybe Int`
+> fmap (+1) $ Just 1
+Just 2
+> fmap (+1) Nothing
+Nothing
+
+-- apply `(> 0) :: Int -> Bool` function
+-- to `Maybe Int`
+-- to get `Maybe Bool`
+> fmap (> 0) $ Just (-1)
+Just False
+> fmap (> 0) $ Just 1
+Just True
+> fmap (> 0) Nothing
+Nothing
+
+-- apply `chr :: Int -> Char` function
+-- to pair `(a, Int)`
+-- to get pair `(a, Char)`
+-- Note that only the second value is being fmap'ed
+> fmap chr (65,65)
+(65,'A')
+> fmap chr ('a',97)
+('a','a')
+```
+
+As you may have guessed, `map` is just a synonym of `fmap` for list. But it can do much more.
+
+With `fmap`, we can reverse strings, convert numbers to strings, perform algebraic and boolean operations, and do other stuff with values inside data types like `[]`, `Maybe`, `Either a`, pair `((,) a)`, and others.
+
+The implementation of `fmap` is data-type-specific. So to use it on a data type, that data type needs to have an implementation for it.
```

Comment:
This seems to be at odds with a statement "you can implement map for an arbitrary type".

Reply:
I've added the related comment [here](https://github.com/serokell/blog-posts/pull/15#discussion_r906841807)

Reply:
I think the problem is that we're showing an unimplementable map'. 

The real type sig is this, right? 

fmap :: Functor f => (a -> b) -> f a -> f b

I think once we introduce this, we can make everything clear. 🤔